### PR TITLE
Offline Setup and Offline Form Display

### PR DESCRIPTION
**Pull request by manuelroemer**
Preparation for offline mode of the forms.

Open issues:

* patient-chart must cache the current visit
* buildmanifest currently contains wrong values for /dist folders due to the renaming of the main.js file

### CODE REVIEW

**Review by FlorianRappl — APPROVED** [review]
LGTM!

**Comment by FlorianRappl** [comment]
Regarding the wrong naming: Let's just not rename "main.js". I think we only would need to change the import map deployer (CI) to use "...main.js" and declare the right one is "package.json". That's it.

File: package.json [hunk 1]
```diff
@@ -29,6 +29,7 @@
     "@angular/platform-browser": "~8.2.14",
     "@angular/platform-browser-dynamic": "~8.2.14",
     "@angular/router": "~8.2.14",
+    "dexie": "^3.0.3",
```
Comment (FlorianRappl) [inline package.json:32]: We still rename "main.js" -> shouldn't we stop doing that?

File: src/openmrs.ts [hunk 2]
```diff
@@ -1,28 +1,52 @@
...
+    "pattern": ".+/ws/rest/v1/obs.+"
+  });
+
+  messageOmrsServiceWorker({
+    "type": "registerDynamicRoute",
+    "pattern": ".+/ws/rest/v1/appui/session.*"
+  });
+
+  subscribeNetworkRequestFailed(({request}) => {
+    console.log('Request failed: ', request);
```
Comment (FlorianRappl) [inline src/openmrs.ts:36]: I guess the `console.log` is temporary here, right?

File: package.json (hunk 1 above)
Comment (manuelroemer) [reply package.json:32]: Yes, I will remove it.

File: src/openmrs.ts (hunk 2 above)
Comment (manuelroemer) [reply src/openmrs.ts:36]: Yep, will remove it.

**Comment by manuelroemer** [comment]
@FlorianRappl Sorry, the push was accidental. I only intended to make a local commit. Muscle memory was stronger though..

File: package.json [hunk 3]
```diff
@@ -10,12 +10,11 @@
     "test-headless": "ng test --watch=false --browsers=ChromeHeadless",
     "lint": "ng lint",
     "e2e": "ng e2e",
-    "build:single-spa-local": "ng build --prod --output-hashing=none --deploy-url https://localhost:4200/ && npm run rename:mainjs",
+    "build:single-spa-local": "ng build --prod --output-hashing=none --deploy-url https://localhost:4200/",
     "build:single-spa-travis": "bash ./scripts/build-travis.sh",
     "serve:single-spa": "ng serve --ssl true  --disable-host-check --deploy-url https://localhost:4202/ --host localhost --port 4202 --live-reload true",
     "serve:dist": "npx angular-http-server -p 4200 --https --cors --path dist/",
-    "build-serve:dist": "npm run build:single-spa-local && npm run serve:dist",
-    "rename:mainjs": "node -e \"require('fs').rename('dist/main-es2015.js', 'dist/openmrs-esm-form-entry-app.js', function(err) { if (err) console.log(err); console.log('File successfully renamed!') })\""
```
Comment (manuelroemer) [inline package.json:18]: With this removed, is there anything else that needs to be updated?

Comment (FlorianRappl) [reply package.json:18]: I don't think so - I think its fine!

Comment (manuelroemer) [reply package.json:18]: Awesome! Then this PR is good to go in my opinion.
It only has a soft dependency on this one: https://github.com/openmrs/openmrs-esm-patient-chart/pull/265

Comment (manuelroemer) [reply package.json:18]: Ah, and the build is failing again. Looking into it.

File: src/offline.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import Dexie, { Table } from "dexie";
-import { NetworkRequestFailedEvent } from "@openmrs/esm-offline";
-import { openmrsFetch } from "@openmrs/esm-api";
+import { NetworkRequestFailedEvent } from "@openmrs/esm-framework";
```
Comment (FlorianRappl) [inline src/offline.ts:2]: Could be a single import, too:

```
import { NetworkRequestFailedEvent, openmrsFetch } from "@openmrs/esm-framework";
```

File: src/app/openmrs-api/openmrs-esm-api.service.ts [hunk 5]
```diff
@@ -1,21 +1,18 @@
 import { Injectable } from '@angular/core';
 import { Observable, of } from 'rxjs';
 import { map } from 'rxjs/operators';
+import {getCurrentUser, openmrsObservableFetch} from '@openmrs/esm-framework';
```
Comment (FlorianRappl) [inline src/app/openmrs-api/openmrs-esm-api.service.ts:4]: Prettier?

**Review by FlorianRappl — COMMENTED** [review]
Yeah looks great!

**Review by FlorianRappl — APPROVED** [review]
Great, thanks! :beers:

**Pull request merged**